contrib/signet/miner

```diff
@@ -0,0 +1,639 @@
+            logging.warning("GBT based off unexpected block (%s not %s), retrying", tmpl["previousblockhash"], bci["bestblockhash"])
+            time.sleep(1)
+            continue
+
+        logging.debug("GBT template: %s", tmpl)
+
+        if tmpl["mintime"] > mine_time:
+            logging.info("Updating block time from %d to %d", mine_time, tmpl["mintime"])
+            mine_time = tmpl["mintime"]
+            if mine_time > now:
+                logging.error("GBT mintime is in the future: %d is %d seconds later than %d", mine_time, (mine_time-now), now)
+                return 1
+
+        # address for reward
+        reward_addr, reward_spk = get_reward_addr_spk(args, tmpl["height"])
+
+        # mine block
+        logging.debug("Mining block delta=%s start=%s mine=%s", seconds_to_hms(mine_time-bestheader["time"]), mine_time, is_mine)
+        mined_blocks += 1
+        psbt = generate_psbt(tmpl, reward_spk, blocktime=mine_time)
+        psbt_signed = json.loads(args.bcli("-stdin", "walletprocesspsbt", input=psbt.encode('utf8')))
+        if not psbt_signed.get("complete",False):
+            logging.debug("Generated PSBT: %s" % (psbt,))
+            sys.stderr.write("PSBT signing failed")
+            return 1
+        block, signet_solution = do_decode_psbt(psbt_signed["psbt"])
+        block = finish_block(block, signet_solution, args.grind_cmd)
+
+        # submit block
+        r = args.bcli("-stdin", "submitblock", input=ToHex(block).encode('utf8'))
+
+        # report
+        bstr = "block" if is_mine else "backup block"
+
+        next_delta = next_block_delta(block.nBits, block.hash, ultimate_target, args.poisson)
+        next_delta += block.nTime - time.time()
+        next_is_mine = next_block_is_mine(block.hash, my_blocks)
+
+        logging.debug("Block hash %s payout to %s", block.hash, reward_addr)
+        logging.info("Mined %s at height %d; next in %s (%s)", bstr, tmpl["height"], seconds_to_hms(next_delta), ("mine" if next_is_mine else "backup"))
+        if r != "":
+            logging.warning("submitblock returned %s for height %d hash %s", r, tmpl["height"], block.hash)
+        lastheader = block.hash
+
+def do_calibrate(args):
+    if args.nbits is not None and args.seconds is not None:
+        sys.stderr.write("Can only specify one of --nbits or --seconds\n")
+        return 1
+    if args.nbits is not None and len(args.nbits) != 8:
+        sys.stderr.write("Must specify 8 hex digits for --nbits")
+        return 1
+
+    TRIALS = 600 # gets variance down pretty low
+    TRIAL_BITS = 0x1e3ea75f # takes about 5m to do 600 trials
+    #TRIAL_BITS = 0x1e7ea75f # XXX
+
+    header = CBlockHeader()
+    header.nBits = TRIAL_BITS
+    targ = nbits_to_target(header.nBits)
+
+    start = time.time()
+    count = 0
+    #CHECKS=[]
+    for i in range(TRIALS):
+        header.nTime = i
+        header.nNonce = 0
+        headhex = header.serialize().hex()
+        cmd = args.grind_cmd.split(" ") + [headhex]
```

Comment:
If `miner calibrate` is called without any additional arguments, there is a stack trace:

```
$ ./miner calibrate 
Traceback (most recent call last):
  File "./miner", line 637, in <module>
    main()
  File "./miner", line 631, in main
    return args.fn(args)
  File "./miner", line 541, in do_calibrate
    cmd = args.grind_cmd.split(" ") + [headhex]
AttributeError: 'NoneType' object has no attribute 'split'
```

One possible way to fix it could be making `--grind-cmd` required for the calibrate command. For `--solvepsbt` and `--generate`, apparently, this is not needed, since `finish_block()` checks for `None` already.

```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -604,8 +604,8 @@ def main():
         sp.add_argument("--address", default=None, type=str, help="Address for block reward payment")
         sp.add_argument("--descriptor", default=None, type=str, help="Descriptor for block reward payment")
 
-    for sp in [solvepsbt, generate, calibrate]:
-        sp.add_argument("--grind-cmd", default=None, type=str, help="Command to grind a block header for proof-of-work")
+    for (sp, is_required) in [(solvepsbt, False), (generate, False), (calibrate, True)]:
+        sp.add_argument("--grind-cmd", default=None, type=str, required=is_required, help="Command to grind a block header for proof-of-work")
 
     args = parser.parse_args(sys.argv[1:])
```

Behaviour after the proposed change:
```
$ ./miner calibrate 
usage: miner calibrate [-h] [--nbits NBITS] [--seconds SECONDS] --grind-cmd GRIND_CMD
miner calibrate: error: the following arguments are required: --grind-cmd
```